python/lsst/daf/persistence/policy.py

```diff
@@ -0,0 +1,392 @@
+#!/usr/bin/env python
+
+#
+# LSST Data Management System
+# Copyright 2015 LSST Corporation.
+#
+# This product includes software developed by the
+# LSST Project (http://www.lsst.org/).
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+#
+
+import collections
+import os
+import UserDict
+import warnings
+import yaml
+
+import lsst.pex.policy as pexPolicy
+
+class Policy(UserDict.UserDict):
+    """Policy implements a datatype that is used by Butler for configuration parameters.
+    It is essentially a dict with key/value pairs, including nested dicts (as values). In fact, it can be
+    initialized with a dict. The only caviat is that keys may NOT contain dots ('.'). This is explained next:
+    Policy extends the dict api so that hierarchical values may be accessed with dot-delimited notiation.
+    That is, foo.getValue('a.b.c') is the same as foo['a']['b']['c'] is the same as foo['a.b.c'], and either
+    of these syntaxes may be used.
+
+    Storage formats supported:
+    - yaml: read and write is supported.
+    - pex policy: read is supported, although this is deprecated and will at some point be removed.
+    """
+
+    defaultPolicyInitData = collections.namedtuple('defaultPolicyInitData', 'productName filePath repos')
+    # Contains data for loading a default policy from a package:
+    # product name: the name of the pacage
+    # filePath: the name of the policy file
+    # the file path from the root of the directory to the dir that contains the file.
+
+    def __init__(self, other=None):
+        """Constructor. Accepts an optional initializer var which may be one of:
+        - a path to a yaml or paf configuration file
```

Comment:
You could possibly make these separate keyword args with defaults of None and avoid a typeswitch (but then you need to define a precedence order in case more than one is specified).